Gh 3

Ross-Alexandra opened the pull request:
This PR should have all the changes necessary for GH-3 to be complete.

Commit:
GH-3 - Added base functionality for GH-3.

Commit:
GH-3 - Fixed failing test.

Commit:
GH-3 - Added unit tests for test_puzzle.

Commit:
GH-3 - Added unit tests for the solver.

Commit:
GH-3 - Fixed bug in tests.

Commit:
GH-3 - Changed prints to logging in verbose.

 - Adde test.

Ross-Alexandra-Review commented on sudoku/puzzle.py:
```diff
@@ -85,6 +98,51 @@ def get_cell_from_box(self, box_position: tuple, cell_position: tuple):
 
+    def get_box_from_cell(self, cell_position: tuple, flatten=False):
+        """ Returns that the box that the cell is in.
+            Don't check for validation as this is just a helper for converting
```
This is no longer true, as we now return the actual box rather than requiring a call like
```self.get_box(get_box_from_cell(0, 0))```.

Ross-Alexandra-Review replied:
Also, need the validation parameter.

Ross-Alexandra replied:
Will update the comment, and add the validation parameter check.

Ross-Alexandra replied:
Resolved in https://github.com/Ross-Alexandra/SudokuPlayer/pull/10/commits/afc6efb7a8610ba2551a72031834661b45a5010d

Ross-Alexandra-Review commented on sudoku/puzzle.py:
```diff
@@ -85,6 +98,51 @@ def get_cell_from_box(self, box_position: tuple, cell_position: tuple):
+        return self.get_box((box_x, box_y), flatten=flatten)
+
+    def is_equal(self, puzzle):
+        """ Don't implement __eq__ as we don't want to implement __hash__.
```
Out of context may not make sense: should include the doc page: https://docs.python.org/3.6/reference/datamodel.html#object.__hash__

Ross-Alexandra replied:
Will add link to docstring.

Ross-Alexandra replied:
Resolved in https://github.com/Ross-Alexandra/SudokuPlayer/pull/10/commits/f08598f50b8de233240466f665899cc5325faf87

Ross-Alexandra-Review commented on sudoku/puzzle.py:
```diff
@@ -85,6 +98,51 @@ def get_cell_from_box(self, box_position: tuple, cell_position: tuple):
+        """
+        return puzzle._puzzle == self._puzzle
+
+    def __str__(self):
```
Should probably include a docstring explaining what the string will look like.

Ross-Alexandra replied:
Will do.

Ross-Alexandra replied:
Resolved in https://github.com/Ross-Alexandra/SudokuPlayer/pull/10/commits/598fe8d1a3a303776fd1727f2457d7d6f6bd6d07

Ross-Alexandra-Review commented on sudoku/puzzle_solver.py:
```diff
@@ -1,3 +1,91 @@
+        self._find_missing_values()
+
+    def _find_missing_values(self):
+        """ Generate and return a dictionary of all the possibilities
```
This method no longer returns.

Ross-Alexandra replied:
Will remove from the docstring.

Ross-Alexandra replied:
Resolved in https://github.com/Ross-Alexandra/SudokuPlayer/pull/10/commits/dbf6676185fea0a7494844d909a6fac8eea6d9df

Ross-Alexandra-Review commented on sudoku/puzzle_solver.py:
```diff
@@ -1,3 +1,91 @@
+        for pattern in self._patterns:
+            pattern.apply_to(self._puzzle)
+
+    def _is_puzzle_solved(self):
```
This should probably also check that the puzzle is valid, that way just filling the thing with 1s doesn't result in a solved puzzle.

Ross-Alexandra replied:
Added a check.

Ross-Alexandra replied:
Resolved by https://github.com/Ross-Alexandra/SudokuPlayer/pull/10/commits/9bc86d5b8a058e6de6eab59f8eb0343b88fbd15c

Ross-Alexandra-Review commented on tests/test_puzzle.py:
```diff
@@ -179,3 +195,23 @@ def test_set_cell_setting_invalid_cell_undoes_move(invalid_val):
+    assert isinstance(str(puzzle), str)
+
+
+def test_is_equal_returns_true_for_equal_puzzles():
```
This test is probably better with only one puzzle, ie
```assert puzzle1.is_equal(puzzle1)```
As this should be guaranteed to be true if this method works.

Ross-Alexandra replied:
Will change.

Ross-Alexandra replied:
Resolved in https://github.com/Ross-Alexandra/SudokuPlayer/pull/10/commits/d712a5578a63aed7b36f488522b7438c6c857c89

Commit:
GH-3 - Updated get_box_from_cell from review.

Commit:
GH-3 - Added reference link to is_equal docstring.
https://docs.python.org/3.6/reference/datamodel.html#object.__hash__

Commit:
GH-3 - Multiple changes

  - Set the default logging level to INFO.
  - Added an example string-ed puzzle to the
    __str__ docstring.

Commit:
GH-3 - Fixed docstring.

Commit:
GH-3 - Added validity check to is puzzle solved.

Commit:
GH-3 - Clairfied a test.
